src/ServiceGetActiveAlarmInfo.c

```diff
@@ -0,0 +1,182 @@
+        motoros2_interfaces__srv__GetActiveAlarmInfo_Request__init(
+            &g_messages_GetActiveAlarmInfo.request),
+        SUBCODE_FAIL_INIT_SERVICE_GET_ACTIVE_ALARM_INFO,
+        "Failed to init request");
+
+    // init response
+    mem_conf_response_.allocator = &g_motoros2_Allocator;
+    mem_conf_response_.rules = mem_rules_response_;
+    mem_conf_response_.n_rules = sizeof(mem_rules_response_) / sizeof(mem_rules_response_[0]);
+    type_support_response_ = ROSIDL_GET_MSG_TYPE_SUPPORT(
+        motoros2_interfaces, srv, GetActiveAlarmInfo_Response);
+
+    motoRosAssert_withMsg(
+        micro_ros_utilities_create_message_memory(
+            type_support_response_, &g_messages_GetActiveAlarmInfo.response, mem_conf_response_),
+        SUBCODE_FAIL_INIT_SERVICE_GET_ACTIVE_ALARM_INFO,
+        "Failed to init response");
+
+    g_messages_GetActiveAlarmInfo.response.alarms.size = 0;
+    g_messages_GetActiveAlarmInfo.response.errors.size = 0;
+
+    rcl_ret_t ret = rclc_service_init_default(
+        &g_serviceGetActiveAlarmInfo, &g_microRosNodeInfo.node,
+        ROSIDL_GET_SRV_TYPE_SUPPORT(motoros2_interfaces, srv, GetActiveAlarmInfo),
+        SERVICE_NAME_GET_ACTIVE_ALARM_INFO);
+    motoRosAssert_withMsg(ret == RCL_RET_OK,
+        SUBCODE_FAIL_INIT_SERVICE_GET_ACTIVE_ALARM_INFO, "Failed to init service (%d)",
+        (int)ret);
+
+    MOTOROS2_MEM_TRACE_REPORT(svc_get_active_alarm_info_init);
+}
+
+void Ros_ServiceGetActiveAlarmInfo_Cleanup()
+{
+    MOTOROS2_MEM_TRACE_START(svc_get_active_alarm_info_fini);
+
+    Ros_Debug_BroadcastMsg("Cleanup service " SERVICE_NAME_GET_ACTIVE_ALARM_INFO);
+    rcl_ret_t ret = rcl_service_fini(
+        &g_serviceGetActiveAlarmInfo, &g_microRosNodeInfo.node);
+    if (ret != RCL_RET_OK)
+        Ros_Debug_BroadcastMsg("Failed cleaning up "
+            SERVICE_NAME_GET_ACTIVE_ALARM_INFO " service: %d", ret);
+
+    bool result = micro_ros_utilities_destroy_message_memory(
+        type_support_response_, &g_messages_GetActiveAlarmInfo.response, mem_conf_response_);
+    Ros_Debug_BroadcastMsg("%s: cleanup response msg memory: %d", __func__, result);
+
+    MOTOROS2_MEM_TRACE_REPORT(svc_get_active_alarm_info_fini);
+}
+
+void Ros_ServiceGetActiveAlarmInfo_Trigger(const void* request_msg, void* response_msg)
+{
+    RCL_UNUSED(request_msg);
+    GetActiveAlarmInfoResponse* response = (GetActiveAlarmInfoResponse*)response_msg;
+
+    Ros_Debug_BroadcastMsg("%s: retrieving active alarm & error info", __func__);
+
+    // retrieve active alarms and/or errors
+    MP_ALARM_CODE_RSP_DATA alarmData;
+    bzero(&alarmData, sizeof(alarmData));
+    if (mpGetAlarmCode(&alarmData) != OK)
+    {
+        // can't continue if M+ API fails
+        Ros_Debug_BroadcastMsg("%s: error retrieving alarm data", __func__);
+        // TODO: use proper error codes + strings
+        rosidl_runtime_c__String__assign(&response->result_message, "M+ API error");
+        response->result_code = -1;
```

Comment:
The `TODO` was mostly a reminder we need to come up with some nice(r) way to deal with return codes. Both here and in other cases.

I could make them part of the `.srv` definition, but wasn't sure that was the best way to go about it.

Reply:
And I'd probably like to define more meaningful error codes, to facilitate automated handling of errors (otherwise clients just see `-1` and would have to interpret the `result_message`).

Reply:
coming back to this:

> I opted to make them part of the response message

this is not a good idea, as it means the `.srv` definition would have to be updated to change or add result codes and strings. This changes the msg hash on Iron and later, which would cause compatibility issues even if none of the actual message fields change.

I'll create a dedicated result codes `.msg`, similar to what we have for other services.